### PR TITLE
core: subRow refactor, rename to subItem

### PR DESCRIPTION
**Pull request by connorjclark**
Continuing from https://github.com/GoogleChrome/lighthouse/pull/10553

This is [option 2](https://github.com/GoogleChrome/lighthouse/pull/10553#issuecomment-635389106) of @brendankenny's suggested changes to sub-rows, including the changes to the renderer to use it.

### CODE REVIEW

File: types/audit-details.d.ts [hunk 1]
```diff
@@ -116,7 +122,8 @@ declare global {
 
       export interface TableItem {
         debugData?: DebugData;
-        [p: string]: undefined | ItemValue | ItemValue[];
+        subRows?: TableSubRows;
```
Comment (connorjclark) [inline types/audit-details.d.ts:125]: Can we just call these `subItems` at this point?

Comment (brendankenny) [reply types/audit-details.d.ts:125]: > Can we just call these subItems at this point?

👍 Makes sense to me, not sure if others were involved in the original bike shedding and will want to weigh in

**Comment by brendankenny** [comment]
@googlebot I consent.

File: types/audit-details.d.ts (hunk 1 above)
Comment (connorjclark) [reply types/audit-details.d.ts:125]: `subRows => subItems` for the detail type, and `subRows => subItemsHeading / subHeading` for the `headings` object?

@patrickhulce @paulirish

File: lighthouse-cli/test/smokehouse/test-definitions/byte-efficiency/expectations.js [hunk 2]
```diff
@@ -155,20 +155,10 @@ const expectations = [
...
-                sourceBytes: [
-                  '4417 +/- 50',
-                  '2200 +/- 50',
-                  '2809 +/- 50',
-                ],
-                sourceWastedBytes: [
-                  '2191 +/- 50',
-                  '2182 +/- 50',
-                  '1259 +/- 50',
+                subItems: [
```
Comment (Beytoven) [inline lighthouse-cli/test/smokehouse/test-definitions/byte-efficiency/expectations.js:158]: This is an excellent improvement!

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 3]
```diff
@@ -315,41 +315,94 @@ class DetailsRenderer {
...
+
+    return rowElem;
+  }
+
+  /**
+   * Renders one or more rows from a details table item.
+   * @param {LH.Audit.Details.OpportunityItem | LH.Audit.Details.TableItem} item
+   * @param {LH.Audit.Details.OpportunityColumnHeading[]} headings
+   */
+  _renderTableItemRows(item, headings) {
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/details-renderer.js:374]: can we come up with some naming options for `_renderTableRow` and `_renderTableItemRows`? :)

I assume `_renderTableItemRows` is saying a table item can be a row or rows, depending on if it had subrows in it? But `_renderTableItem` also sounds like a singular thing (like `_renderTableValue`), not a row or rows...

`_renderTableRowsFromItem`?

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 4]
```diff
@@ -315,41 +315,94 @@ class DetailsRenderer {
...
+      if (heading && valueElement) {
+        const classes = `lh-table-column--${heading.valueType}`;
+        this._dom.createChildOf(rowElem, 'td', classes).appendChild(valueElement);
+      } else {
+        // Empty cell is rendered for a column if:
+        // - the pair is null
+        // - the heading key is null
+        // - the value is undefined/null
+        this._dom.createChildOf(rowElem, 'td', 'lh-table-column--empty');
+      }
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/details-renderer.js:363]: to be able to skip the `heading && ...` checks, what about breaking the empty cell check in two so the no heading case can be ruled out early? Something like
```js
for (const heading of headings) {
  // Empty cell if no heading or heading key for this column.
  if (!heading || !heading.key) {
    this._dom.createChildOf(rowElem, 'td', 'lh-table-column--empty');
    continue;
  }

  const value = item[heading.key];
  const valueElement = this._renderTableValue(value, heading);

  if (valueElement) {
    const classes = `lh-table-column--${heading.valueType}`;
    this._dom.createChildOf(rowElem, 'td', classes).appendChild(valueElement);
  } else {
    // Empty cell if there's no value to put in it.
    this._dom.createChildOf(rowElem, 'td', 'lh-table-column--empty');
  }
}
```

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 5]
```diff
@@ -315,41 +315,94 @@ class DetailsRenderer {
...
    */
   _getCanonicalizedHeading(heading) {
-    let subRows;
-    if (heading.subRows) {
+    let subHeading;
+    if (heading.subHeading) {
       // @ts-ignore: It's ok that there is no text.
-      subRows = this._getCanonicalizedHeading(heading.subRows);
-      if (!subRows.key) {
+      subHeading = this._getCanonicalizedHeading(heading.subHeading);
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/details-renderer.js]: Best I could do was introduce two new functions, one for canonicalizing subHeading and another for deriving a header from a subHeading/parent.

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 6]
```diff
@@ -315,41 +315,94 @@ class DetailsRenderer {
...
+   * Renders one or more rows from a details table item.
+   * @param {LH.Audit.Details.OpportunityItem | LH.Audit.Details.TableItem} item
+   * @param {LH.Audit.Details.OpportunityColumnHeading[]} headings
+   */
+  _renderTableItemRows(item, headings) {
+    const fragment = this._dom.createFragment();
+    fragment.append(this._renderTableRow(item, headings));
+
+    // A single details item can expand into multiple table rows. These additional table rows
+    // are called sub-rows.
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/details-renderer.js]: I missed updating this. and this comment was more useful here when the function was bigger and more complex. moved it to the function jsdoc.

File: types/audit-details.d.ts [hunk 7]
```diff
@@ -85,15 +91,15 @@ declare global {
...
+      type ItemValue = string | number | boolean | DebugData | NodeValue | SourceLocationValue | LinkValue | UrlValue | CodeValue | TableSubItems;
 
       // TODO: drop TableColumnHeading, rename OpportunityColumnHeading to TableColumnHeading and
       // use that for all table-like audit details.
 
       export interface TableColumnHeading {
         /**
          * The name of the property within items being described.
-         * If null, subRows must be defined, and the first sub-row will be empty.
+         * If null, subHeading must be defined, and the first sub-row will be empty.
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply types/audit-details.d.ts]: let's do `subItem row` in this comment.

and `sub-item-row` in css.

File: lighthouse-core/test/report/html/renderer/details-renderer-test.js [hunk 8]
```diff
@@ -570,76 +570,98 @@ describe('DetailsRenderer', () => {
...
       it('renders', () => {
         const details = {
           type: 'table',
-          headings: [{key: 'url', itemType: 'url', subRows: {key: 'sources', itemType: 'code'}}],
+          headings: [{key: 'url', itemType: 'url', subHeading: {key: 'source', itemType: 'code'}}],
           items: [
-            {url: 'https://www.example.com', sources: ['a', 'b', 'c']},
+            {
+              url: 'https://www.example.com',
+              subItems: makeSubitems([
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/test/report/html/renderer/details-renderer-test.js:588]: I'm trying to reduce object fatigue to increase test readability. The function is free and avoids an extra level of nesting, which for these `details` objects can be overwhelming.

File: lighthouse-core/report/html/renderer/details-renderer.js (hunk 3 above)
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/details-renderer.js:413]: `_renderTableRowsFromItem` sgtm!

File: types/audit-details.d.ts [hunk 9]
```diff
@@ -85,15 +91,15 @@ declare global {
...
+      type ItemValue = string | number | boolean | DebugData | NodeValue | SourceLocationValue | LinkValue | UrlValue | CodeValue | TableSubItems;
 
       // TODO: drop TableColumnHeading, rename OpportunityColumnHeading to TableColumnHeading and
       // use that for all table-like audit details.
 
       export interface TableColumnHeading {
         /**
          * The name of the property within items being described.
-         * If null, subRows must be defined, and the first sub-row will be empty.
+         * If null, subHeading must be defined, and the first subItem row will be empty.
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply types/audit-details.d.ts]: tried to improve the comment.

writing these details renderer tests is cumbersome... I wish we just diffed HTML snapshots here :/

File: lighthouse-cli/test/smokehouse/test-definitions/byte-efficiency/expectations.js (hunk 2 above)
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/test-definitions/byte-efficiency/expectations.js:158]: supposed to be `subItems: {items: [...]}`, I think?

**Comment by connorjclark** [comment]
One last thing from me:

Should `subHeading` be named `subItemsHeading`?

**Comment by connorjclark** [comment]
oops

**Comment by brendankenny** [comment]
oops?

**Comment by connorjclark** [comment]
I had an outstanding discussion item and didn't realize the merge when green label was here when I fixed the tests. Nbd.

**Pull request merged**